### PR TITLE
BUG - Fastlane Unit Tests #173

### LINKED ISSUE #173: BUG - Fastlane Unit Tests
## Describe the bug
* Currently the Unit tests lane for `Fastlane` is not running, all tests appear as passed even if some failed

## To Reproduce
1. Write a test that failed under `Basic-Car-Maintenance-Tests`
2. Run the command `bundle exec fastlane unit_tests`
3. The results in the command line doesn't show any failed test.

## Expected behavior
If there's a failed test, the command line results should look like below:

## Screenshots
Refer to the screenshot above

## Device Info (please complete the following information):
N/A

## Additional Context
* This is tied to the Github Action to run unit tests on PR to `dev` branch

### PR DESCRIPTION
**Pull request by anupdsouza**
# What it Does
* Closes #173 
* Provides coverage report #71 using xcov

# How I Tested
* Executed unit tests locally using fastlane, `bundle exec fastlane unit_tests`
* Ran github workflow after pushing to branch

# Notes
* Removed `build_for_testing` config setting which was only building the code but not executing the tests. [Ref fastlane docs.](https://docs.fastlane.tools/actions/run_tests/).
* Updated gemfile to install `xcov` gem for code coverage.
* Added `xcov` action in fastfile to provide code coverage.
* Minimum code coverage has not been specified yet as there aren't enough tests written. This can be updated later by adding `minimum_coverage_percentage: 80.0` with an acceptable benchmark value to the `xcov` action in the fastfile.
* Added `.xcovignore` file in fastlane folder. This can be used to exclude files we do not need to test or require coverage to be gathered for. I believe model/business logic files are more suited for testing. This can be updated as development progresses.
* Added `xcodebuild_formatter` config in fastlane to use `xcbeautify` for cleaner logs.
* Updated github workflow to install dependencies via bundle (for gems) & via homebrew (for xcbeautify)
* Right now the plan only executes unit tests via the `only_testing` config flag in fastlane. If removed, it will also execute UI tests ( #54 ) but I've observed the workflow taking over an hour to complete on GitHub. So maybe it's something we need to weigh the pros & cons for?
* Updated workflow such that only the last initiated workflow executes in order to prevent multiple workflows from running on multiple code push actions. See `concurrency` in workflow file.
* In order to run locally, make sure to first run `bundle update` for installing gem dependencies and then run `brew install xcbeautify` to install xcbeautify for cleaner logging. You might need to use `sudo` incase you run into errors for write permissions to the ruby directory

# Screenshots ~ Local
Screenshot 2023-10-20 at 5 26 24 PM

Screenshot 2023-10-20 at 5 26 48 PM

Screenshot 2023-10-20 at 5 27 48 PM

Coverage report ~ `fastlane/CodeCoverageReport/index.html`
---
Screenshot 2023-10-20 at 5 28 18 PM

# Screenshots ~ Github workflow
---
Failing tests
---
Screenshot 2023-10-20 at 10 03 20 AM

Screenshot 2023-10-20 at 10 03 31 AM

Passing tests
---
Screenshot 2023-10-20 at 9 48 38 AM

Screenshot 2023-10-20 at 1 06 11 PM

# Miscellaneous
* We can add a unit test status badge to the ReadMe as well to provide a quick status of the unit testing workflow. It'd look something like this Unit Tests Workflow Badge but it would say passing or failing depending on the unit test workflow status. If you feel this adds value, please feel free to create an enhancement or improvement & assign to me & I'd be happy to take it up!

### CODE REVIEW

**Comment by mikaelacaron** [comment]
Thanks for the detailed PR!!
> Right now the plan only executes unit tests via the only_testing config flag in fastlane. If removed, it will also execute UI tests ( https://github.com/mikaelacaron/Basic-Car-Maintenance/issues/54 ) but I've observed the workflow taking over an hour to complete on GitHub. So maybe it's something we need to weigh the pros & cons for?

Yeah that's definitely an issue, let's not run them at this moment, and I may just run them manually for now (when it's needed)

> We can add a unit test status badge to the ReadMe as well to provide a quick status of the unit testing workflow. It'd look something like this Unit Tests Workflow Badge but it would say passing or failing depending on the unit test workflow status. If you feel this adds value, please feel free to create an enhancement or improvement & assign to me & I'd be happy to take it up!

Yes I'll do this after I merge your PR!

I'll look at this soon, I haven't gone through the code changes, and I need to run it all locally still

File: fastlane/Fastfile [hunk 1]
```diff
@@ -1,30 +1,35 @@
...
-#
-#     https://docs.fastlane.tools/actions
-#
-# For a list of all available plugins, check out
-#
-#     https://docs.fastlane.tools/plugins/available-plugins
-#
-
-# Uncomment the line if you want fastlane to automatically update itself
-# update_fastlane
```
Comment (mikaelacaron) [inline fastlane/Fastfile]: Please keep this here! I'm not super familiar with fastlane, and this is still useful

File: fastlane/Fastfile [hunk 2]
```diff
@@ -1,30 +1,35 @@
...
+			skip_build: true,
+			xcodebuild_formatter: 'xcbeautify -qq --is-ci --renderer github-actions',
+			output_types: 'junit',
+			output_directory: "./fastlane/UnitTestsReport"
+		)
+
+		xcov(
+			scheme: "Basic-Car-Maintenance",
+			only_project_targets: true,
+			output_directory: "./fastlane/CodeCoverageReport"
```
Comment (mikaelacaron) [inline fastlane/Fastfile:46]: Does this need to be ignored?

**Review by mikaelacaron — CHANGES_REQUESTED** [review]
> Minimum code coverage has not been specified yet as there aren't enough tests written. This can be updated later by adding minimum_coverage_percentage: 80.0 with an acceptable benchmark value to the xcov action in the fastfile.

Can you add this in the right spot, but comment it out?

I have a few questions too, noted below

I haven't been able to run this locally, `bundle update` has been timing out

File: fastlane/Fastfile (hunk 1 above)
Comment (anupdsouza) [reply fastlane/Fastfile]: Done.

File: fastlane/Fastfile (hunk 2 above)
Comment (anupdsouza) [reply fastlane/Fastfile:46]: Not sure what you mean? `xcov` is the fastlane action that runs coverage for the project.

File: fastlane/Fastfile (hunk 1 above)
Comment (anupdsouza) [reply fastlane/Fastfile]: Done.

File: fastlane/Fastfile (hunk 2 above)
Comment (anupdsouza) [reply fastlane/Fastfile:46]: Updated gitignore file to exclude reports from accidentally being committed to the repo

**Review by mikaelacaron — APPROVED** [review]
Great work!! I got it to run locally! Not sure what happened the first time

Thanks for your patience!

**Pull request merged**